### DISCUSSION TITLE
Reconstruction of experimental data with a polychromatic x-ray source

### DISCUSSION THREAD

**Discussion by figueroa1395** [astra-toolbox/astra-toolbox#429] [Q&A] [answered]
Hello all!

I have read in a previous discussion how to simulate the case of a polychromatic x-ray source, however I believe my case is slightly different. I have the projections' data of an unknown object (hence I do not have any information about internal structure or attenuation coefficients) obtained from an experimental setup with a polychromatic x-ray source, and, as I read in the other discussion, astra comes hard-coded with the assumption of a monochromatic source. Hence, is there a way to take into account that the source in the experimental setup is not monochromatic and I do not know the target's internal structure? I want to reduce artifacts such as beam hardening. Moreover, I can have access to the source spectrum via the detector response and I can perform other experimental measurements, but I am just not sure about how to deal with this, or if it even makes sense.

As a bonus question (Let me know if a separate thread is better), does astra support the use of multiple gpus natively? If not, is there a known workaround to enable support?

Thanks!

**Comment by askorikov** [comment, marked as answer]
Hi! Beam hardening correction is a rich topic, and there are many methods to do it that may be more or less successful depending on the situation and differ in how computationally demanding they are. Most of these methods attempt to iteratively compute reconstruction and estimate attenuation coefficients of the materials within. Astra Toolbox doesn't provide such a functionality, but you can use it as a building block to implement any of the existing methods.

Multi-GPU support is implemented for forward- and backprojection operations, as illustrated in https://github.com/astra-toolbox/astra-toolbox/blob/master/samples/python/s020_3d_multiGPU.py

**Reply by figueroa1395** [reply]
Thanks!

**Comment by stephoog** [comment]
Hey! You could try using [LEAP](https://github.com/llnl/leap). This library can do polychromatic projections and beam hardening correction. The beam hardening stuff uses a separate library called [XrayPhysics](https://github.com/kylechampley/XrayPhysics) which you could probably use with ASTRA if you want.
